common/djangoapps/philu_commands/management/commands/data_consistency_check_onboarding_optimization.py

```diff
@@ -0,0 +1,114 @@
+
+    if expected_value != actual_value:
+        log.error(ERROR_MSG_MISMATCH.format(field_name='INTERESTS',
+                                            profile_id=user_extended_profile.id,
+                                            user_id=user_extended_profile.user_id,
+                                            expected_val=expected_value,
+                                            actual_val=actual_value))
+
+
+def compare_learners_related(user_extended_profile):
+    expected_value = list([k for k, _ in INTERESTED_LEARNERS if getattr(user_extended_profile, k)])
+    actual_value = list(user_extended_profile.learners_related)
+
+    if expected_value != actual_value:
+        log.error(ERROR_MSG_MISMATCH.format(field_name='LEARNERS_RELATED',
+                                            profile_id=user_extended_profile.id,
+                                            user_id=user_extended_profile.user_id,
+                                            expected_val=expected_value,
+                                            actual_val=actual_value))
+
+
+def compare_goals(user_extended_profile):
+    expected_value = list([k for k, _ in GOALS if getattr(user_extended_profile, k)])
+    actual_value = list(user_extended_profile.goals)
+
+    if expected_value != actual_value:
+        log.error(ERROR_MSG_MISMATCH.format(field_name='GOALS',
+                                            profile_id=user_extended_profile.id,
+                                            user_id=user_extended_profile.user_id,
+                                            expected_val=expected_value,
+                                            actual_val=actual_value))
+
+
+def compare_hear_about_philanthropyu(user_extended_profile):
+    expected_value = get_expected_hear_about_philanthropyu(user_extended_profile)
+
+    hear_about_philanthropyu_list = list(user_extended_profile.hear_about_philanthropyu)
+    actual_value = hear_about_philanthropyu_list[0] if len(hear_about_philanthropyu_list) > 0 else ''
+
+    if expected_value != actual_value:
+        log.error(ERROR_MSG_MISMATCH.format(field_name='HEAR_ABOUT_PHILANTHROPYU',
+                                           profile_id=user_extended_profile.id,
+                                           user_id=user_extended_profile.user_id,
+                                           expected_val=expected_value,
+                                           actual_val=actual_value))
+
+
+def get_expected_hear_about_philanthropyu(user_extended_profile):
+    hear_about_philanthropy = user_extended_profile.hear_about_philanthropy
+    hear_about_philanthropy_other = user_extended_profile.hear_about_philanthropy_other
+
+    if hear_about_philanthropy is None:
+        return ''
+
+    if hear_about_philanthropy == 'Other':
+        return hear_about_philanthropy_other
+
+    for key, value in HEAR_ABOUT_PHILANTHROPY_DICT.items():
+        if hear_about_philanthropy == value:
+            return key
+
```

Comment:
Here also, the case is similar to the previous one, except for one addition. First of all, if such a record appears in hear_about_philanthropy that is neither NULL, neither 'Other', nor belongs to HEAR_ABOUT_PHILANTHROPY_DICT, such a record needs to be taken care of before running the migration. If I return '' in the end, it would be logically incorrect, because it would mean that expected value against an **invalid** record in hear_about_philanthropy is '', which is not the case. Such an invalid record needs to be taken care of, while writing the migration query.

Again, I did make sure, before writing the query, that such a case does not appear. For this, I had the following query run on production:

```
SELECT id, hear_about_philanthropy, hear_about_philanthropy_other
FROM onboarding_userextendedprofile
WHERE hear_about_philanthropy IS NOT NULL AND hear_about_philanthropy NOT IN ('A Philanthropy University Partner (Global Giving, +Acumen or another)',
									  										  'A Colleague From My Organization',
									  										  'A Friend Or Colleague (Not From My Organization)',
									  										  'An Internet Search',
									  										  'A LinkedIn Advertisement',
									  										  'A Facebook Advertisement',
									  										  'Twitter (Not From A Colleague)',
									  										  'Other');
```

The result of this query was an empty set, thereby ensuring no invalid record, or a record with '' exists in this field in the production db.